### DISCUSSION TITLE
Couple of minor improvements

### DISCUSSION THREAD

**Discussion by tlinhart** [snok/asgi-correlation-id#69] [Ideas]
Hi @sondrelg I wonder if you would be willing to accept another two minor improvements.

The default value for the correlation ID is `None` which is perfectly fine. However, `None` also appears in logs when there is no correlation ID set. People might want to alter this behaviour and log something more neutral as `None` looks too Pythonic. E.g. some symbol like `-` (which is what for example nginx logs when variable is not defined). This could be achieved by introducing a new parameter to the log filter constructor which would be then supplied as the default value to the call to `correlation_id.get()`. The parameter could be named for example `default_value` (makes sense in the context of a log filter) or `none_replacement`.

The second thing relates to the case when incoming correlation ID does not pass validations and is thus replaced with a generated value. It logs a warning which reads
```
Generated new request ID (xxxxx), since request header value failed validation
```
but `xxxxx` contains the original header value, not the newly generated one. Given the wording of the warning, it's a bit confusing as one would expect that the log would contain the newly generated value. I'd either suggest to replace the behaviour and log the newly generated value, or change the warning to
```
Generated new request ID (xxxxx-new), since request header value (xxxxx-original) failed validation
```
The later approach might be a better solution as it would allow you to correlate both IDs, the original (which might come from and be logged by a downstream service) and the new one (which might be propagated to and logged by an upstream service).

What do you think?

**Comment by sondrelg** [comment]
So for the first suggestion, you are proposing this change, correct?

```diff
class CorrelationIdFilter(Filter):
    """Logging filter to attached correlation IDs to log records"""

-    def __init__(self, name: str = '', uuid_length: Optional[int] = None):
+    def __init__(self, name: str = '', uuid_length: Optional[int] = None, default_value: Optional[str] = None):
        super().__init__(name=name)
        self.uuid_length = uuid_length
+     self.default_value = default_value

    def filter(self, record: 'LogRecord') -> bool:
-        cid = correlation_id.get()
+        cid = correlation_id.get() or default_value
        record.correlation_id = _trim_string(cid, self.uuid_length)
        return True
```

That seems reasonable :+1: 

For the second, perhaps logging the logger, without change, after we call `correlation_id.set(id_value)` would make most sense? That way the request is aggregated with remaining request when queried by the new request ID, but the log still only contains the strictly necessary information. Do you see any flaws in that approach?

**Reply by tlinhart** [reply]
As for the first suggestion, yeah, that's the idea. I just vote for using
```
cid = correlation_id.get(default_value)
```
instead of
```
cid = correlation_id.get() or default_value
```
It's IMHO semantically cleaner (as per [the docs](https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar.get)) and also accounts for the case should you ever change the default value of the context variable from `None` to e.g. an empty string. The same change would be done to both log filters, i.e. to `CeleryTracingIdsFilter` as well, just to be complete. I'll prepare a PR for this one if you agree.

As for the second suggestion, my point was two-fold. First, as per the current wording it's a bit confusing when after stating about a new request ID you specify the old one (in parentheses). That could be solved by your suggestion, if I understand it correctly. Or, the warning could be rephrased to e.g.
```
Generated new request ID, since request header value (xxxxx) failed validation
```
That would still log only the original header but in a more understandable way. This all just really depends on what your goal really was – log the original value or the newly generated one? My second suggestion was to log both, but as per the comment from @JonasKs there can be a security risk.

**Reply by sondrelg** [reply]
Both sound good to me, pending input from Jonas :slightly_smiling_face:

**Comment by JonasKs** [comment]
First request: 👍 I agree 😊 

Second request: I don't agree, not because it's not a nice idea, but because of security concerns. Sonarqube, bandit etc. will fail if this is implemented, and we've seen the catastrophic impact usage of headers, without parsing can have with log4j. I specifically removed it in Django-GUID due to this: https://github.com/snok/django-guid/pull/69

If you really want this, I suggest creating a custom validator where you log the input, I don't think it should be the default of the library.

**Reply by tlinhart** [reply]
I don't quite understand now. Well I do understand there can be a security risk in logging HTTP header values. But that's what you are doing right now, don't you? Logging the [original header value](https://github.com/snok/asgi-correlation-id/blob/a8008bfc05f209c7736eca0536c7aa975c9ae429/asgi_correlation_id/middleware.py#L63) (that failed your validation).

**Reply by sondrelg** [reply]
Could you elaborate on this @JonasKs?

**Reply by JonasKs** [reply]
> But that's what you are doing right now, don't you? Logging the [original header value](https://github.com/snok/asgi-correlation-id/blob/a8008bfc05f209c7736eca0536c7aa975c9ae429/asgi_correlation_id/middleware.py#L63) (that failed your validation).

Ah, shit, I didn't even know! I thought I had checked this library when I fixed  `Django-GUID`. Sorry for the confusion. 

In short, I do not think we should log header values, it's considered a security risk, and it got a lot of traction after log4j.   
Scanners like Bandit and Sonarqube will fail if they see logging of incoming headers, if those headers don't pass validation.

**Reply by tlinhart** [reply]
I understand and vote for security first approach :+1: On the other hand, if there's no way how to identify the individual request that got affected (incoming correlation ID value replaced with a new one), than the warning log message feels a bit superfluous. There are three posibilities as I see it:
1. Get rid of the warning completely.
2. Leave the warning message but remove any header values from the message.
3. Replace the original header value (as is logged at the moment) with the newly generated one. I find don't find this a security risk because you are the one who creates the value and is thus under your control (as opposed to the original header value which you have no control over).

**Reply by sondrelg** [reply]
Why is logging the old one a security risk?

Yeah I vote for 3, assuming we don't want to log the old one.

**Reply by tlinhart** [reply]
From what I understand, it's because it's something that comes from outside, something you don't have control over. And as it's a (custom) HTTP header, it can be easily forged.

**Reply by sondrelg** [reply]
So like sql injection, but with logs? How are logs evaluated in a way that that's a security concern?

**Reply by JonasKs** [reply]
Yes, and if there is a security issue with the logging package (like in Log4j), e.g. the log interpolation evaluates to something it shouldn't, it can be bad. 

Log4j is basically: Send something to the app, the app logs it (using log4j), but it had a vulnerability, so you could send links, scripts etc. into it, and it would download it and execute it. In other words, if the `logging` lib has a buffer overflow vulnerability triggered by a string with 9991 emojis, all applications logging headers would be affected. Log4j all over. 

So, after log4j, the consensus is: Do not log headers that has not been validated.

**Reply by JonasKs** [reply]
https://tldr.engineering/tldr-log4j-vulnerability/

**Reply by tlinhart** [reply]
I don't know the details, the Log4j case is discussed for example [here](https://www.swarmnetics.com/blog/apache-log4j-vulnerability-explained/). But I'd say it doesn't have to necessarily impact the system that logs the messages, it can also affect other systems. Imagine when you aggregate the logs with ELK stack or Loki. In the end, it stores the messages in some sort of storage so it potentially can lead to something like SQL injection.

**Reply by JonasKs** [reply]
Yeah, I have no good memories from the log4j times personally 😁 So much was vulnerable..

**Reply by sondrelg** [reply]
I think it's pretty useful to know what value failed validation though. Having no trace of it seems impractical :thinking: After log4j I would expect there to be a lot less likelihood of a repeat occurrence of this issue in the Python logging standard library :upside_down_face: I'll think about it over the weekend :slightly_smiling_face: 

In the meantime, the filter change seems uncontroversial :+1:

**Reply by JonasKs** [reply]
Less likely to do the same lookup-logic, yeah, but buffer overflows and all other kinds of vulnerabilities can still happen. I don't mind having it as a setting, but I personally don't want this library to log unparsed incoming input. I have my own middlewares (or route handlers) to log incoming values I care about.

**Reply by JonasKs** [reply]
Overflows are the [second most common vulnerability](https://www.cvedetails.com/vulnerability-list/vendor_id-10210/product_id-18230/opov-1/Python-Python.html) in Python, and log interpolation isn't safe from it. 

Log being aggregated to other systems also face the same issues, as @tlinhart pointed out.

**Reply by tlinhart** [reply]
I get your point @JonasKs. To be true, I've never thought about this being a security risk until now. And as the whole or at least a major point of this middleware is to be able to log the request ID coming from an input HTTP request header, I now I fully appreciate the existence of `validator` and the default checking for an UUID value.

**Reply by tlinhart** [reply]
Anyway, at the moment, the original header value is being logged if it doesn't pass the validations. Which I assume is exactly the case of a potential security issue. How about that? As pointed out [above](https://github.com/snok/asgi-correlation-id/discussions/69#discussioncomment-5099323), I see three options where the third one is IMHO a good compromise. The newly generated header value is something you can control (thus ensure validity) and gives you at least something at hand when correlating the logged requests.

**Reply by JonasKs** [reply]
Sounds like a good solution to me. 😊

**Reply by JonasKs** [reply]
> So like sql injection, but with logs? How are logs evaluated in a way that that's a security concern?

I didn't see this reply, sorry. I think I still answered it indirectly? 🤔 

As a side note: I recommend reading [this article](https://fasterthanli.me/articles/the-bottom-emoji-breaks-rust-analyzer) about how an Emoji broke rust-analyzer, it is a good article which kind of explains the obscure things that can happen with user input.

**Comment by tlinhart** [comment]
I've created PRs for both discussed changes, please take a look.
